src/index.ts

```diff
@@ -3,64 +3,88 @@
-    const png = canvases[i].toDataURL('image/png');
-    (
-      canvasesCopy[i].parentElement as Element
-    ).innerHTML = `<img src="${png}" />`;
+    canvasesCopy[i].replaceWith(canvasToHtml(canvases[i]));
   }
-
-  let svgs = dom.querySelectorAll('svg');
-  let svgsCopy = domCopy.querySelectorAll('svg');
-
-  const promises = [];
-
+  const promises: Promise<void>[] = [];
+  const svgs = dom.querySelectorAll('svg');
+  const svgsCopy = domCopy.querySelectorAll('svg');
   for (let i = 0; i < svgs.length; i++) {
-    const svgDOM = svgs[i];
-    const svgDOMCopy = svgsCopy[i];
-    const width = svgDOM.clientWidth;
-    const height = svgDOM.clientHeight;
-    const svgString = svgDOM.outerHTML;
-    const canvas = document.createElement('canvas');
-    canvas.setAttribute('width', width.toString());
-    canvas.setAttribute('height', height.toString());
-    const ctx = canvas.getContext('2d');
-    if (ctx) {
-      ctx.fillStyle = 'white';
-      ctx.fillRect(0, 0, canvas.width, canvas.height);
-    }
-    const image = new Image();
-    const svg = new Blob([svgString], {
-      type: 'image/svg+xml;charset=utf-8',
-    });
-    const url = URL.createObjectURL(svg);
+    promises.push(
+      svgToHtml(svgs[i], (result) => {
+        svgsCopy[i].replaceWith(result);
+      }),
+    );
+  }
 
+  const imgs = dom.querySelectorAll('img');
+  const imgsCopy = domCopy.querySelectorAll('img');
+  for (let i = 0; i < imgs.length; i++) {
     promises.push(
-      new Promise<void>((resolve) => {
-        image.onload = () => {
-          if (ctx) {
-            ctx.drawImage(image, 0, 0);
-          }
-          const png = canvas.toDataURL('image/png');
-          const img = document.createElement('img');
-          img.src = png;
-          svgDOMCopy.replaceWith(img);
-          URL.revokeObjectURL(url);
-          resolve();
-        };
+      imgToHtml(imgs[i], (result) => {
+        imgsCopy[i].replaceWith(result);
       }),
     );
-    image.src = url;
   }
 
   await Promise.all(promises);
   return domCopy.innerHTML;
 }
+
+function canvasToHtml(canvas: HTMLCanvasElement) {
+  const url = canvas.toDataURL('image/png');
+  const img = new Image();
+  img.src = url;
+  return img;
+}
+async function svgToHtml(
+  svg: SVGSVGElement,
+  callback: (result: HTMLImageElement) => void,
+) {
+  const width = svg.clientWidth;
+  const height = svg.clientHeight;
+  const base64 = btoa(new XMLSerializer().serializeToString(svg));
+  const url = `data:image/svg+xml;base64,${base64}`;
+  const image = new Image();
+  return new Promise<void>((resolve) => {
+    image.onload = () => {
+      const canvas = document.createElement('canvas');
+      canvas.width = width;
+      canvas.height = height;
+      const ctx = canvas.getContext('2d') as CanvasRenderingContext2D;
+      ctx.drawImage(image, 0, 0, width, height);
+      callback(canvasToHtml(canvas));
+      resolve();
+    };
+    image.src = url;
+  });
+}
+async function imgToHtml(
```

Comment:
Could you check that is works if the image is already base64 encoded ?

So if the html contains for example https://codepen.io/jamiekane/pen/YayWOa

I have the impression that fetch may fail and actually it should not try to replace this encoded image.

Reply:
All results are saved in src/tests/domToHtml.test.tsx-snapshots

Reply:
> I have the impression that fetch may fail and actually it should not try to replace this encoded image.

I'm actually replacing the domCopy elements with the encoded images